### ISSUE TITLE
Unexpected carriage return in `exec` output

### ISSUE THREAD

**Issue by rcgoodfellow** [oxidecomputer/falcon#79]
The following `exec` command is [used in a4x2](https://github.com/oxidecomputer/omicron/blob/1c27e88e12ee2aaca864934b50828251f947c468/.github/buildomat/jobs/a4x2-deploy.sh#L183-L184) environments to collect address information from a VM.

```bash
customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local'")
```

This results in an unexpected carriage return in the output as evidenced by

```bash
echo "'$customer_edge_addr'"
192.168.1.213'
```

A workaround for this is cleaning up with `tr`.

```bash
customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local'" | tr -d '\r')
```

which then yields the expected result

```bash
echo "'$customer_edge_addr'"
'192.168.1.213'
```

**Comment by elaine-oxide** [comment]
Workaround needs refinement.

```
$ customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local'" | tr -d '\r')

$ echo $customer_edge_addr
172.20.2.188

$ echo "'$customer_edge_addr'"
'172.20.2.188'

// different run but same commands above
$ echo $customer_edge_addr | cat -v
^[[?2004l172.20.2.160

**Comment by dancrossnyc** [comment]
Instead of `tr -d '\r'`, perhaps, `tr -d '[^0-9.]'`

It's sort of odd to me that terminal escape codes are being emitted into the output; they're almost certainly coming from `jq`, though. Perhaps run `jq -rMj`, or even pipe `jq`'s output through `cat` (in the pipeline run by `a4x2`).

I'm curious if this solves it, though:

```
$ customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -rMj '.[0].addr_info[] | select(.dynamic == true) | .local'")
```

**Comment by dancrossnyc** [comment]
More generally, I suspect the unwanted artifacts in the output here are a result of the `ip` and `jq` commands being run in a VM that has, importantly, allocated a terminal for them and interposed a line discipline on the stream that does things like convert newlines to newline/carriage return pairs and so on. You could preface them with an `stty` command to suppress that behavior, but is there a reason not to run `jq` back on the host? E.g., something like:

```
customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
```

I haven't tested it, but I suspect that running `jq` locally would eliminate the problem.

**Comment by elaine-oxide** [comment]
> Instead of `tr -d '\r'`, perhaps, `tr -cd '[0-9.]'`
> 
> It's sort of odd to me that terminal escape codes are being emitted into the output; they're almost certainly coming from `jq`, though. Perhaps run `jq -rMj`, or even pipe `jq`'s output through `cat` (in the pipeline run by `a4x2`).
> 
> I'm curious if this solves it, though:
> 
> ```
> $ customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -rMj '.[0].addr_info[] | select(.dynamic == true) | .local'")
> ```

This variation does not solve it.

```
$ customer_edge_addr=$(./a4x2 exec ce "ip -4 -j addr show enp0s10 | jq -rMj '.[0].addr_info[] | select(.dynamic == true) | .local'")

$ echo $customer_edge_addr
172.20.2.169

$ echo "'$customer_edge_addr'"
172.20.2.169'

$ echo $customer_edge_addr | cat -v
^[[?2004l^M172.20.2.169
```

**Comment by elaine-oxide** [comment]
> More generally, I suspect the unwanted artifacts in the output here are a result of the `ip` and `jq` commands being run in a VM that has, importantly, allocated a terminal for them and interposed a line discipline on the stream that does things like convert newlines to newline/carriage return pairs and so on. You could preface them with an `stty` command to suppress that behavior, but is there a reason not to run `jq` back on the host? E.g., something like:
> 
> ```
> customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
> ```
> 
> I haven't tested it, but I suspect that running `jq` locally would eliminate the problem.

This variation has some kind of problem.

```
$ customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
parse error: Invalid numeric literal at line 1, column 2

$ ./a4x2 exec ce 'ip -4 -j addr show enp0s10'
[{"ifindex":4,"ifname":"enp0s10","flags":["BROADCAST","MULTICAST","UP","LOWER_UP"],"mtu":1500,"qdisc":"pfifo_fast","operstate":"UP","group":"default","txqlen":1000,"addr_info":[{"family":"inet","local":"172.20.2.169","prefixlen":24,"broadcast":"172.20.2.255","scope":"global","dynamic":true,"label":"enp0s10","valid_life_time":2185,"preferred_life_time":2185},{"family":"inet","local":"192.168.20.100","prefixlen":24,"broadcast":"192.168.20.255","scope":"global","label":"enp0s10","valid_life_time":4294967295,"preferred_life_time":4294967295}]}]
```

```
root@ce:~# ip -4 -j addr show enp0s10
[{"ifindex":4,"ifname":"enp0s10","flags":["BROADCAST","MULTICAST","UP","LOWER_UP"],"mtu":1500,"qdisc":"pfifo_fast","operstate":"UP","group":"default","txqlen":1000,"addr_info":[{"family":"inet","local":"172.20.2.169","prefixlen":24,"broadcast":"172.20.2.255","scope":"global","dynamic":true,"label":"enp0s10","valid_life_time":3448,"preferred_life_time":3448},{"family":"inet","local":"192.168.20.100","prefixlen":24,"broadcast":"192.168.20.255","scope":"global","label":"enp0s10","valid_life_time":4294967295,"preferred_life_time":4294967295}]}]
```

**Comment by dancrossnyc** [comment]
> > More generally, I suspect the unwanted artifacts in the output here are a result of the `ip` and `jq` commands being run in a VM that has, importantly, allocated a terminal for them and interposed a line discipline on the stream that does things like convert newlines to newline/carriage return pairs and so on. You could preface them with an `stty` command to suppress that behavior, but is there a reason not to run `jq` back on the host? E.g., something like:
> > ```
> > customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I haven't tested it, but I suspect that running `jq` locally would eliminate the problem.
> 
> This variation has some kind of problem.
> 
> ```
> $ customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
> parse error: Invalid numeric literal at line 1, column 2

My suspicion is that this has to do with terminal goo getting into the output stream, because of the TTY issue.

> $ ./a4x2 exec ce 'ip -4 -j addr show enp0s10'
> [{"ifindex":4,"ifname":"enp0s10","flags":["BROADCAST","MULTICAST","UP","LOWER_UP"],"mtu":1500,"qdisc":"pfifo_fast","operstate":"UP","group":"default","txqlen":1000,"addr_info":[{"family":"inet","local":"172.20.2.169","prefixlen":24,"broadcast":"172.20.2.255","scope":"global","dynamic":true,"label":"enp0s10","valid_life_time":2185,"preferred_life_time":2185},{"family":"inet","local":"192.168.20.100","prefixlen":24,"broadcast":"192.168.20.255","scope":"global","label":"enp0s10","valid_life_time":4294967295,"preferred_life_time":4294967295}]}]
> ```

It would be interesting to see what happens when you pipe the output of the `a4x2 exec ce` command through `cat -v`.

> ```
> root@ce:~# ip -4 -j addr show enp0s10
> [{"ifindex":4,"ifname":"enp0s10","flags":["BROADCAST","MULTICAST","UP","LOWER_UP"],"mtu":1500,"qdisc":"pfifo_fast","operstate":"UP","group":"default","txqlen":1000,"addr_info":[{"family":"inet","local":"172.20.2.169","prefixlen":24,"broadcast":"172.20.2.255","scope":"global","dynamic":true,"label":"enp0s10","valid_life_time":3448,"preferred_life_time":3448},{"family":"inet","local":"192.168.20.100","prefixlen":24,"broadcast":"192.168.20.255","scope":"global","label":"enp0s10","valid_life_time":4294967295,"preferred_life_time":4294967295}]}]
> ```

That looks fine; if I copy and paste it into a file locally, the output is normal:

```
: doctor; cat foo.json
[{"ifindex":4,"ifname":"enp0s10","flags":["BROADCAST","MULTICAST","UP","LOWER_UP"],"mtu":1500,"qdisc":"pfifo_fast","operstate":"UP","group":"default","txqlen":1000,"addr_info":[{"family":"inet","local":"172.20.2.169","prefixlen":24,"broadcast":"172.20.2.255","scope":"global","dynamic":true,"label":"enp0s10","valid_life_time":3448,"preferred_life_time":3448},{"family":"inet","local":"192.168.20.100","prefixlen":24,"broadcast":"192.168.20.255","scope":"global","label":"enp0s10","valid_life_time":4294967295,"preferred_life_time":4294967295}]}]
: doctor; jq -r '.[0].addr_info[] | select(.dynamic == true) | .local' < foo.json | cat -v
172.20.2.169
: doctor; x=$(jq -r '.[0].addr_info[] | select(.dynamic == true) | .local' < foo.json)
: doctor; echo $x
172.20.2.169
: doctor; echo $x | cat -v
172.20.2.169
: doctor;
```

Something you could try is prefacing the `ip` command with `stty raw`; as in:

```
customer_edge_addr=$(./a4x2 exec ce 'stty raw; ip -4 -j addr show enp0s10' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')
```

**Comment by dancrossnyc** [comment]
Elaine and I just talked about this, and the error is definitely due to terminal output coming out of the `a4x2` command (really, it's coming from Linux, just via the `a4x2 exec`). A hokey way to work around it is to cut that out with sed:

```
./a4x2 exec ce 'ip -4 -j addr show enp0s10' | sed 's/^........//' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local'
```

**Comment by elaine-oxide** [comment]
I did a full test including finishing the final steps of a4x2 that use this variable, and confirmed that the latest workaround works.

```
$ customer_edge_addr=$(./a4x2 exec ce 'ip -4 -j addr show enp0s10' | sed 's/^........//' | jq -r '.[0].addr_info[] | select(.dynamic == true) | .local')

$ echo $customer_edge_addr
172.20.2.169

$ echo "'$customer_edge_addr'"
'172.20.2.169'

$ echo $customer_edge_addr | cat -v
172.20.2.169
```

**Comment by rcgoodfellow** [comment]
fwiw the serial command code is here

https://github.com/oxidecomputer/falcon/blob/84b8147e03290fa96d8693e5b668e6d032542907/lib/src/serial.rs#L194-L232

**Comment by dancrossnyc** [comment]
It's independent of how data from the guest is read, unfortunately. The issue is really the terminal codes coming from the Linux side. In particular, I believe this is `bash` dumping terminal escape codes into the output stream.

The data code that @elaine-oxide was seeing coming from the Linux side are due to handling bracketed paste mode (something like `^[[2004l^M` I think; sorry, I didn't copy it down, but that's the escape sequence for "disable bracketed paste mode"). That's almost certainly coming from bash, or rather its `readline` library, when it detects that it's running as the login shell on a terminal device. Unfortunately, short of invoking the `exec` command outside of the context of a TTY, `bash` will always make that assumption: there's no good way for it to know that it's talking to something other than a user in front a terminal on the far end of a (virtual) serial port.

However, it just occurred to me that there may be another way to deal with this, which is just to turn off bracketed paste mode in bash in the serial login code (which already assumes a bash-y kind of shell anyway). Sending, `bind 'set enable-bracketed-paste off'`, perhaps right before or after setting the terminal type, will probably disable it.

**Comment by dancrossnyc** [comment]
falcon#80 might address this.

**Comment by elaine-oxide** [comment]
Fixed in #80.
